blog/2016-02-15-ceylon-browser-again.md

```diff
@@ -0,0 +1,222 @@
+the filesystem, you can also avoid the HTTP server and load everything for the filesystem.
+
+## Using browser APIs
+
+Now that we have bootstrapped a Ceylon application running in a browser, it's time to do actual things
+that leverage browser APIs. To do this, we'll use the brand new `ceylon.interop.browser` which was
+introduced in the Ceylon SDK 1.2.1 a few days ago. Basically, it's a set of 
+[`dynamic` interfaces](/blog/2014/10/13/dynamic-interfaces/) that
+allow wrapping native JS objects returned by the browser in nice typed Ceylon instances. For example, this
+interface represents the browser's `Document`:
+
+<!-- try: -->
+    shared dynamic Document satisfies Node & GlobalEventHandlers {
+        shared formal String \iURL;
+        shared formal String documentURI;
+        ...
+        shared formal HTMLCollection getElementsByTagName(String localName);
+        shared formal HTMLCollection getElementsByClassName(String classNames);
+        ...
+    }
+
+An instance of `Document` can be retrieved via the toplevel object `document`, just like in JavaScript:
+
+<!-- try: -->
+    shared Document document => window.document;
+
+Note that `window` is also a toplevel instance of the dynamic interface `Window`.
+
+`ceylon.interop.browser` contains lots of interfaces related to:
+
+* [DOM nodes](https://www.w3.org/TR/dom/#nodes)
+* [DOM events](https://www.w3.org/TR/dom/#events)
+* [DOM traversal](https://www.w3.org/TR/dom/#traversal)
+* [XMLHttpRequest](https://www.w3.org/TR/XMLHttpRequest/)
+
+Making an AJAX call, retrieving the result and adding it to a `<div>` is now super easy in Ceylon:
+
+<!-- try: -->
+    import ceylon.interop.browser.dom {
+        document,
+        Event
+    }
+    import ceylon.interop.browser {
+        newXMLHttpRequest
+    }
+
+    shared void run() {
+        value req = newXMLHttpRequest();
+
+        req.onload = void (Event evt) {
+            if (exists container = document.getElementById("container")) {
+                value title = document.createElement("h1");
+                title.textContent = "Hello from Ceylon";
+                container.appendChild(title);
+                
+                value content = document.createElement("p");
+                content.innerHTML = req.responseText;
+                container.appendChild(content);
+            }
+        };
+        
+        req.open("GET", "/msg.txt");
+        req.send();
+    }
+
+## Going further
+
+Dynamic interfaces are really nice when it comes to using JavaScript objects in Ceylon. They are somewhat
+similar to TypeScript's [type definitions](https://github.com/DefinitelyTyped/DefinitelyTyped), which means
+in theory, it is possible to use any JavaScript framework directly from Ceylon, provided that someone writes dynamic
+interfaces for its API.
+
+The Ceylon team is currently looking for ways to [convert TypeScript type definitions to dynamic interfaces](https://github.com/ceylon/ceylon/issues/3041),
```

Comment:
Well, technically, we won’t convert TypeScript definitions to dynamic interfaces – they’ll look similar from a user’s perspective, but there’s a lot in TypeScript that dynamic interfaces alone can’t represent (toplevel members and functions, classes with constructors, a different calling convention). Perhaps this would be better:

> The Ceylon team is currently looking for ways to *load TypeScript definitions and make them available to Ceylon modules*

or something like that.